### DISCUSSION TITLE
Ready firmware to Engine Hat

### DISCUSSION THREAD

**Discussion by Shelis1976** [hatlabs/discussions#88] [Q&A]
Can Anybody help/share main.cpp to the Engine Hat ready to use?
I uploaded the excample and thought that transmition into can nmea2000 is into the firmware but there is not.
Im not super good on this thats why I'm asking for help. 
Please help

Thanks in advance 
Tom

**Comment by jacktheflyer** [comment]
Have you the Hat or the HALMET?

**Reply by Shelis1976** [reply]
HAT

**Reply by jacktheflyer** [reply]
I have an Arduino sketch which I developed for the SH-ESP32. You are welcome to try it But it is very untidy.

**Reply by Shelis1976** [reply]
Please share.
Tom

**Reply by jacktheflyer** [reply]
see attached.  History to help you wade through it.  
This has been cobbled together to allow for OTA updates.  Also caters for two engines each with four temperatures, oil pressure, RPM.  You will also see two tank levels, three voltages and an BME380 for environment.  Thank levels and BME380 are working on another board and are different to what you see here. (can send it).  This works on the SH-ESP32 for the Port engine.  My intention was to just change over the NMEA sentences at line-1036 from Port to Stbd.  Should work but haven't yet tried this particular version on Stbd yet.
If you have any suggestion how to share it better pls let me know.
If you have any suggestions at all pls feel free to let me know.
Good luck wading through it.
rgds
John
[NMEA2000-DataSender_rev16.3_bluelight.ino.txt](https://github.com/hatlabs/discussions/files/13933403/NMEA2000-DataSender_rev16.3_bluelight.ino.txt)

**Comment by newbezz** [comment]
I have the Helmet. When I clone from github everything works fine. No error (two warnings i yellow, but I think these are old ones). The compiled app uploads to Halmet, which displays "Host: Halmet" on the OLED, and the two green RX and TX LED flickers. BUT there is no trace of the devise on my NMEA test rig. This has a Maretron DMS250 display, a B&G plotter and an actisense NGT-1 gateway. These 3 devises show each other presence on the N2k bus, but no trace of the Halmet. 
I have recompiled, recompiled, and recompiled the SW, rearranged and rearranged the physical N2k network, but all to no avail. 
Last year I went through the same with two SH-ESP32s, but thought they were bad and gave up. I have no problem in setting up a Teensy 3.2 running direcly on Timos original N2k library, and this thing shows up on the N2k bus. 
HELP - I am getting mad ~~~))) 
Best
Jorgen

**Reply by jacktheflyer** [reply]
i was having a lot of issues with the serial port of the HALMET until I closed the CAN termination.  All fell into place after that.

**Reply by mairas** [reply]
@newbezz I have a nagging suspicion: did you configure the WiFi? If not, and especially if the display is only showing "Host: halmet" and nothing else, it's likely it's waiting for the WiFi to be configured. This shouldn't be mandatory -- I'll figure out a workaround for it right away.

**Reply by mairas** [reply]
I have now updated the example firmware to allow disabling Signal K output. This is useful in pure-N2K setups. Note that as opposed to my previous message statement, WiFi still needs to be configured before the device is usable. Device configuration would be impossible otherwise.

**Reply by jacktheflyer** [reply]
Matti
I didn't configure the WiFi in the example program but my program is working on SH-ESP32 for OTA updates.  Have sent a separate email to you as I started to thin one of the boards was faulty.  Would you prefer I posted here?
rgds
John
